File: Submissions/680_Valid_Palindrome_II.cpp
class Solution {
public:
    bool validPalindrome(string s) {
        int n = s.size();
        for(int i=0; i<n/2; ++i){
            if(s[i] == s[n-1-i]) continue;
            
            return isValid(s, i+1, n-1-i) || isValid(s, i, n-2-i);
        }
        
        return true;
    }
    
    bool isValid(string s, int start, int end){
        int i = start, j = end;
        while(i<j){
            if(s[i]!=s[j]) return false;
            ++i, --j;
        }
        return true;
    }
};

File: Kotlin/680_Valid_Palindrome_II.kt
class Solution {
    fun validPalindrome(s: String): Boolean {
        val n = s.length
        for(i in 0 until n / 2) {
            if(s[i] == s[n - i - 1]) continue

            return isPalindrome(s, i + 1, n - i - 1) || isPalindrome(s, i, n - i - 2)
        }

        return true
    }
    
    fun isPalindrome(s: String, begin: Int, end: Int): Boolean {
        var i = begin; var j = end
        while(i < j) {
            if (s[i++] != s[j--]) return false
        }
        
        return true
    }
}